File: internal/command/info.go
package command

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"time"

	"github.com/adamavenir/fray/internal/core"
	"github.com/adamavenir/fray/internal/db"
	"github.com/adamavenir/fray/internal/types"
	"github.com/spf13/cobra"
)

type agentInfo struct {
	GUID         string   `json:"guid"`
	AgentID      string   `json:"agent_id"`
	Status       string   `json:"status"`
	Purpose      string   `json:"purpose"`
	Nicks        []string `json:"nicks"`
	RegisteredAt string   `json:"registered_at"`
	LastSeen     string   `json:"last_seen"`
	LeftAt       *string  `json:"left_at"`
	MessageCount int64    `json:"message_count"`
	ClaimCount   int64    `json:"claim_count"`
}

type channelInfo struct {
	Initialized  bool              `json:"initialized"`
	ChannelID    string            `json:"channel_id,omitempty"`
	ChannelName  string            `json:"channel_name,omitempty"`
	Path         string            `json:"path,omitempty"`
	CreatedAt    string            `json:"created_at,omitempty"`
	LastActivity string            `json:"last_activity,omitempty"`
	MessageCount int64             `json:"message_count,omitempty"`
	AgentCount   int               `json:"agent_count,omitempty"`
	Config       map[string]string `json:"config,omitempty"`
	Agents       []agentInfo       `json:"agents,omitempty"`
}

// NewInfoCmd creates the info command.
func NewInfoCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "info",
		Short: "Show channel information and roster",
		RunE: func(cmd *cobra.Command, args []string) error {
			showGlobal, _ := cmd.Flags().GetBool("global")
			jsonMode, _ := cmd.Flags().GetBool("json")

			if showGlobal {
				config, err := core.ReadGlobalConfig()
				if err != nil {
					return writeCommandError(cmd, err)
				}
				if config == nil || len(config.Channels) == 0 {
					if jsonMode {
						return json.NewEncoder(cmd.OutOrStdout()).Encode(map[string]any{"channels": []channelInfo{}})
					}
					fmt.Fprintln(cmd.OutOrStdout(), "No channels registered")
					return nil
				}

				channels := make([]channelInfo, 0, len(config.Channels))
				for _, channel := range config.Channels {
					info := getChannelInfo(channel.Path)
					if info.Initialized {
						channels = append(channels, info)
					}
				}

				if jsonMode {
					return json.NewEncoder(cmd.OutOrStdout()).Encode(map[string]any{"channels": channels})
				}

				fmt.Fprintf(cmd.OutOrStdout(), "Channels (%d):\n\n", len(channels))
				for _, info := range channels {
					formatChannelInfo(cmd.OutOrStdout(), info, true)
					fmt.Fprintln(cmd.OutOrStdout())
				}
				return nil
			}

			project, err := core.DiscoverProject("")
			if err != nil {
				if jsonMode {
					return json.NewEncoder(cmd.OutOrStdout()).Encode(channelInfo{Initialized: false})
				}
				fmt.Fprintln(cmd.OutOrStdout(), "Not initialized")
				fmt.Fprintln(cmd.OutOrStdout(), "Run: fray init")
				return nil
			}

			info := getChannelInfo(project.Root)
			if !info.Initialized {
				if jsonMode {
					return json.NewEncoder(cmd.OutOrStdout()).Encode(channelInfo{Initialized: false})
				}
				fmt.Fprintln(cmd.OutOrStdout(), "Not initialized")
				fmt.Fprintln(cmd.OutOrStdout(), "Run: fray init")
				return nil
			}

			if jsonMode {
				return json.NewEncoder(cmd.OutOrStdout()).Encode(info)
			}
			formatChannelInfo(cmd.OutOrStdout(), info, false)
			return nil
		},
	}

	cmd.Flags().Bool("global", false, "show info for all registered channels")
	return cmd
}

func getChannelInfo(projectRoot string) channelInfo {
	frayDir := filepath.Join(projectRoot, ".fray")
	if _, err := os.Stat(frayDir); err != nil {
		return channelInfo{Initialized: false}
	}

	dbPath := filepath.Join(frayDir, "fray.db")
	if _, err := os.Stat(dbPath); err != nil {
		return channelInfo{Initialized: false}
	}

	project := core.Project{Root: projectRoot, DBPath: dbPath}
	dbConn, err := db.OpenDatabase(project)
	if err != nil {
		return channelInfo{Initialized: false}
	}
	defer dbConn.Close()
	if err := db.InitSchema(dbConn); err != nil {
		return channelInfo{Initialized: false}
	}

	entries, err := db.GetAllConfig(dbConn)
	if err != nil {
		return channelInfo{Initialized: false}
	}
	config := map[string]string{}
	for _, entry := range entries {
		config[entry.Key] = entry.Value
	}

	agents, err := db.GetAllAgents(dbConn)
	if err != nil {
		return channelInfo{Initialized: false}
	}
	claimCounts, err := db.GetClaimCountsByAgent(dbConn)
	if err != nil {
		return channelInfo{Initialized: false}
	}
	messageCounts, err := getMessageCounts(dbConn)
	if err != nil {
		return channelInfo{Initialized: false}
	}

	totalMessages, err := getTotalMessageCount(dbConn)
	if err != nil {
		return channelInfo{Initialized: false}
	}

	lastTS, err := getLastMessageTS(dbConn)
	if err != nil {
		return channelInfo{Initialized: false}
	}

	projectConfig, _ := db.ReadProjectConfig(dbPath)
	createdAt := ""
	if projectConfig != nil {
		createdAt = projectConfig.CreatedAt
	}

	infos := make([]agentInfo, 0, len(agents))
	for _, agent := range agents {
		infos = append(infos, toAgentInfo(agent, messageCounts, claimCounts, projectConfig))
	}

	sort.Slice(infos, func(i, j int) bool {
		return infos[i].LastSeen > infos[j].LastSeen
	})

	info := channelInfo{
		Initialized:  true,
		ChannelID:    config["channel_id"],
		ChannelName:  config["channel_name"],
		Path:         projectRoot,
		CreatedAt:    createdAt,
		MessageCount: totalMessages,
		AgentCount:   len(agents),
		Config:       config,
		Agents:       infos,
	}
	if lastTS != nil {
		info.LastActivity = time.Unix(*lastTS, 0).UTC().Format(time.RFC3339)
	}
	return info
}

func toAgentInfo(agent types.Agent, messageCounts map[string]int64, claimCounts map[string]int64, config *db.ProjectConfig) agentInfo {
	registered := time.Unix(agent.RegisteredAt, 0).UTC().Format(time.RFC3339)
	lastSeen := time.Unix(agent.LastSeen, 0).UTC().Format(time.RFC3339)
	var leftAt *string
	if agent.LeftAt != nil {
		value := time.Unix(*agent.LeftAt, 0).UTC().Format(time.RFC3339)
		leftAt = &value
	}
	return agentInfo{
		GUID:         agent.GUID,
		AgentID:      agent.AgentID,
		Status:       normalizeOptionalValue(agent.Status),
		Purpose:      normalizeOptionalValue(agent.Purpose),
		Nicks:        agentNicksForGUID(config, agent.GUID),
		RegisteredAt: registered,
		LastSeen:     lastSeen,
		LeftAt:       leftAt,
		MessageCount: messageCounts[agent.AgentID],
		ClaimCount:   claimCounts[agent.AgentID],
	}
}

func formatChannelInfo(out io.Writer, info channelInfo, showPath bool) {
	if !info.Initialized {
		fmt.Fprintln(out, "Not initialized")
		fmt.Fprintln(out, "Run: fray init")
		return
	}

	fmt.Fprintf(out, "Channel: %s (%s)\n", info.ChannelName, info.ChannelID)
	if showPath && info.Path != "" {
		fmt.Fprintf(out, "  path: %s\n", info.Path)
	}
	if info.CreatedAt != "" {
		createdTs, err := time.Parse(time.RFC3339, info.CreatedAt)
		if err == nil {
			fmt.Fprintf(out, "  created: %s\n", formatRelative(createdTs.Unix()))
		}
	}
	if info.LastActivity != "" {
		lastTs, err := time.Parse(time.RFC3339, info.LastActivity)
		if err == nil {
			fmt.Fprintf(out, "  last activity: %s\n", formatRelative(lastTs.Unix()))
		}
	}
	fmt.Fprintf(out, "  messages: %d\n", info.MessageCount)
	fmt.Fprintf(out, "  agents: %d\n", info.AgentCount)

	if info.Config != nil {
		custom := map[string]string{}
		for key, value := range info.Config {
			if key == "channel_id" || key == "channel_name" {
				continue
			}
			custom[key] = value
		}
		if len(custom) > 0 {
			fmt.Fprintln(out, "  config:")
			keys := make([]string, 0, len(custom))
			for key := range custom {
				keys = append(keys, key)
			}
			sort.Strings(keys)
			for _, key := range keys {
				fmt.Fprintf(out, "    %s: %s\n", key, custom[key])
			}
		}
	}

	if len(info.Agents) > 0 {
		fmt.Fprintf(out, "  roster (%d):\n", len(info.Agents))
		for _, agent := range info.Agents {
			status := ""
			if agent.LeftAt != nil {
				status = " (left)"
			}
			lastTs, err := time.Parse(time.RFC3339, agent.LastSeen)
			if err != nil {
				continue
			}
			label := formatAgentLabel(agent.AgentID, agent.Nicks)
			fmt.Fprintf(out, "    %s%s\n", label, status)
			fmt.Fprintf(out, "      status: %s\n", formatOptionalString(agent.Status))
			fmt.Fprintf(out, "      purpose: %s\n", formatOptionalString(agent.Purpose))
			fmt.Fprintf(out, "      last seen: %s\n", formatRelative(lastTs.Unix()))
			fmt.Fprintf(out, "      messages: %d\n", agent.MessageCount)
		}
	}
}

File: src/commands/info.ts
import { Command } from 'commander';
import fs from 'fs';
import path from 'path';
import Database from 'better-sqlite3';
import { discoverProject } from '../core/project.js';
import { readGlobalConfig } from '../core/config.js';
import { initSchema } from '../db/schema.js';
import { readProjectConfig } from '../db/jsonl.js';
import { getAllAgents, getAllConfig, getClaimCountsByAgent } from '../db/queries.js';
import { formatRelative } from '../core/time.js';
import type { Agent, ConfigEntry } from '../types.js';

interface AgentInfo {
  guid: string;
  agent_id: string;
  goal: string | null;
  bio: string | null;
  registered_at: string;
  last_seen: string;
  left_at: string | null;
  message_count: number;
  claim_count: number;
}

interface ChannelInfo {
  initialized: boolean;
  channel_id?: string;
  channel_name?: string;
  path?: string;
  created_at?: string;
  last_activity?: string;
  message_count?: number;
  agent_count?: number;
  config?: Record<string, string>;
  agents?: AgentInfo[];
}

function getChannelInfo(dbPath: string, projectRoot: string): ChannelInfo | null {
  const mmDir = path.join(projectRoot, '.mm');
  if (!fs.existsSync(mmDir)) {
    return null;
  }

  const fullDbPath = path.join(mmDir, 'mm.db');
  if (!fs.existsSync(fullDbPath)) {
    return null;
  }

  try {
    const db = new Database(fullDbPath);
    initSchema(db);

    // Get config
    const configEntries = getAllConfig(db);
    const config: Record<string, string> = {};
    for (const entry of configEntries) {
      config[entry.key] = entry.value;
    }

    // Get agents with counts
    const agents = getAllAgents(db);
    const claimCounts = getClaimCountsByAgent(db);

    const messageCounts = new Map<string, number>();
    const messageCountRows = db.prepare(`
      SELECT from_agent as agent_id, COUNT(*) as count
      FROM mm_messages
      GROUP BY from_agent
    `).all() as { agent_id: string; count: number }[];
    for (const row of messageCountRows) {
      messageCounts.set(row.agent_id, row.count);
    }

    // Get total message count
    const totalMessages = db.prepare('SELECT COUNT(*) as count FROM mm_messages').get() as { count: number };

    // Get last activity
    const lastMessage = db.prepare('SELECT MAX(ts) as ts FROM mm_messages').get() as { ts: number | null };

    // Get project config for created_at
    const projectConfig = readProjectConfig(fullDbPath);

    const agentInfos: AgentInfo[] = agents.map(agent => ({
      guid: agent.guid,
      agent_id: agent.agent_id,
      goal: agent.goal,
      bio: agent.bio,
      registered_at: new Date(agent.registered_at * 1000).toISOString(),
      last_seen: new Date(agent.last_seen * 1000).toISOString(),
      left_at: agent.left_at ? new Date(agent.left_at * 1000).toISOString() : null,
      message_count: messageCounts.get(agent.agent_id) || 0,
      claim_count: claimCounts.get(agent.agent_id) || 0,
    }));

    // Sort agents by last_seen descending
    agentInfos.sort((a, b) =>
      new Date(b.last_seen).getTime() - new Date(a.last_seen).getTime()
    );

    db.close();

    return {
      initialized: true,
      channel_id: config.channel_id,
      channel_name: config.channel_name,
      path: projectRoot,
      created_at: projectConfig?.created_at,
      last_activity: lastMessage.ts ? new Date(lastMessage.ts * 1000).toISOString() : undefined,
      message_count: totalMessages.count,
      agent_count: agents.length,
      config,
      agents: agentInfos,
    };
  } catch {
    return null;
  }
}

function formatChannelDisplay(info: ChannelInfo, showPath: boolean = false): void {
  if (!info.initialized) {
    console.log('Not initialized');
    console.log('Run: mm init');
    return;
  }

  console.log(`Channel: ${info.channel_name} (${info.channel_id})`);
  if (showPath && info.path) {
    console.log(`  path: ${info.path}`);
  }
  if (info.created_at) {
    const createdTs = Math.floor(new Date(info.created_at).getTime() / 1000);
    console.log(`  created: ${formatRelative(createdTs)}`);
  }
  if (info.last_activity) {
    const lastTs = Math.floor(new Date(info.last_activity).getTime() / 1000);
    console.log(`  last activity: ${formatRelative(lastTs)}`);
  }
  console.log(`  messages: ${info.message_count}`);
  console.log(`  agents: ${info.agent_count}`);

  // Show non-standard config
  if (info.config) {
    const standardKeys = new Set(['channel_id', 'channel_name']);
    const customConfig = Object.entries(info.config).filter(([k]) => !standardKeys.has(k));
    if (customConfig.length > 0) {
      console.log('  config:');
      for (const [key, value] of customConfig) {
        console.log(`    ${key}: ${value}`);
      }
    }
  }

  if (info.agents && info.agents.length > 0) {
    console.log(`  roster (${info.agents.length}):`);
    for (const agent of info.agents) {
      const status = agent.left_at ? ' (left)' : '';
      const lastSeenTs = Math.floor(new Date(agent.last_seen).getTime() / 1000);
      console.log(`    @${agent.agent_id}${status} - ${agent.message_count} msgs, last seen ${formatRelative(lastSeenTs)}`);
    }
  }
}

export function infoCommand(): Command {
  return new Command('info')
    .description('Show channel information and roster')
    .option('--global', 'show info for all registered channels')
    .action(async (options, cmd) => {
      const jsonMode = cmd.optsWithGlobals().json || false;
      const showGlobal = options.global || false;

      if (showGlobal) {
        const globalConfig = readGlobalConfig();

        if (!globalConfig || Object.keys(globalConfig.channels).length === 0) {
          if (jsonMode) {
            console.log(JSON.stringify({ channels: [] }, null, 2));
          } else {
            console.log('No channels registered');
          }
          return;
        }

        const channels: ChannelInfo[] = [];

        for (const [channelId, channel] of Object.entries(globalConfig.channels)) {
          const info = getChannelInfo(path.join(channel.path, '.mm', 'mm.db'), channel.path);
          if (info) {
            channels.push(info);
          }
        }

        if (jsonMode) {
          console.log(JSON.stringify({ channels }, null, 2));
        } else {
          console.log(`Channels (${channels.length}):\n`);
          for (const info of channels) {
            formatChannelDisplay(info, true);
            console.log('');
          }
        }
      } else {
        // Local channel info
        let project;
        try {
          project = discoverProject();
        } catch {
          if (jsonMode) {
            console.log(JSON.stringify({ initialized: false }, null, 2));
          } else {
            console.log('Not initialized');
            console.log('Run: mm init');
          }
          return;
        }

        const info = getChannelInfo(path.join(project.root, '.mm', 'mm.db'), project.root);

        if (!info) {
          if (jsonMode) {
            console.log(JSON.stringify({ initialized: false }, null, 2));
          } else {
            console.log('Not initialized');
            console.log('Run: mm init');
          }
          return;
        }

        if (jsonMode) {
          console.log(JSON.stringify(info, null, 2));
        } else {
          formatChannelDisplay(info);
        }
      }
    });
}
